manual/io_custom_classes/index.md

```diff
@@ -204,6 +204,44 @@ For more information on ROOT files, see [ROOT files]({{ '/manual/root_files' | r
 For more information on TTree, see [Trees]({{ '/manual/trees' | relative_url }}).
 
 
+## The `ClassDef` macro
+
+The `ClassDef` macro can be inserted in a class definition to add some reflection capabilities to it. It also attaches a "version number" to the class that can be used for [schema evolution]({{ '/manual/io' | relative_url }}).
+
+Having a `ClassDef` is mandatory for classes inheriting from `TObject`, otherwise it is an optional ROOT I/O performance optimization.
+
+The syntax is:
+
+{% highlight C++ %}
+ClassDef(<ClassName>, <VersionNumber>)
+{% endhighlight %}
+
+The version number identifies this particular version of the class. A version number equal to 0 tells ROOT to not store the class in a root file, but only its base classes (if any).
+
+`ClassDef` injects some methods in the class definition useful for runtime reflection.
+Here are the most important ones:
+
+- `static const char *Class_Name()`: returns the class name as a C-string
+- `static TClass *Class()`: returns a [`TClass`](https://root.cern/doc/master/classTClass.html) instance that can be used to query information about the class
+- `MAYBE_VIRTUAL TClass *IsA() const MAYBE_OVERRIDE`: same as `Class()`, but it returns the `TClass` corresponding to the concrete type in case of a pointer to a base
+- `MAYBE_VIRTUAL void ShowMembers(TMemberInspector &insp) const MAYBE_OVERRIDE`: useful to query the list of members of a class at runtime (see [`TMemberInspector`](https://root.cern/doc/master/classTMemberInspector.html))
+
+Use `ClassDefOverride` to include the `override` keyword in the appropriate injected methods.<br>
+Use `ClassDefNV` to not mark any of the injected methods as `virtual`.
+
+_**Example**_
+
+{% highlight C++ %}
+class MyClass {
+   // A good initial version number is 3
```

Comment:
Looks arbitrary. I guess there is an explanation and we don't want to expose it? A sentence above says `A version number equal to 0 tells ROOT to not store the class in a root file`, but no explanation for version 1 or 2

Reply:
@Axel-Naumann mentioned when writing the document that 3 is actually a _better_ initial version number than others. We need his help here.

Reply:
Here is the full reason for this seemingly arbitrary number:

A class layout can be either versioned (using a `ClassDef` or `#pragma link C++ options=version(3) class MyClass` or `<class name="pat::Muon"  ClassVersion="12">` or the macro `RootClassVersion`) or unversioned (in which case the class layout checksum is stored and used instead).

However when an unversioned class layout is loaded (either via the dictionary or via `TStreamerInfo`s for a file), they are given a version number starting at 1.

So if a class at any point in its development life unversioned and later becomes versioned and the user want to be able to load "older" files that used the unversioned class layout, the user need to leave some 'space' in the `TStreamerInfo` array (indexed on the version number) for the unversioned class layout.

Consequently the "real" advice is to start the versioning at "number of distinct unversioned class layout" that you might need to load at once (in a process that has this class now versioned).  

To "keep it simple", we came up with a compromise (the value 3) that was "safe enough" (Can read at the same time old files from 2 distinct eras of the life of that class) while not wasting space (if no files with unversioned class layout are read those first few array slot are "wasted" memory usage).

Reply:
```suggestion
   // Note that the initial version number be greater than the number of previously, unnumbered (i.e. lacking a
   // explicit `ClassDef`) versions of the class. If unsure, `3` is typically a good compromise.
```